### PR TITLE
Class declarations should be block scoped.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/Microsoft/TypeScript/issues/2578
Fixes https://github.com/Microsoft/TypeScript/issues/2579

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -9757,7 +9757,12 @@ module ts {
         function checkClassDeclaration(node: ClassDeclaration) {
             // Grammar checking
             if (node.parent.kind !== SyntaxKind.ModuleBlock && node.parent.kind !== SyntaxKind.SourceFile) {
-                grammarErrorOnNode(node, Diagnostics.class_declarations_are_only_supported_directly_inside_a_module_or_as_a_top_level_declaration);
+                grammarErrorOnFirstToken(node, Diagnostics.class_declarations_are_only_supported_directly_inside_a_module_or_as_a_top_level_declaration);
+            }
+
+            // node.flags & NodeFlags.Default || kind === SyntaxKind.ClassExpression
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Turn this into a TODO.

Also, for future reference, I am more in favor of `ClassyDeclaration` over `ClassLikeDeclaration`.

File: tests/baselines/reference/anonymousClassExpression1.errors.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+tests/cases/compiler/anonymousClassExpression1.ts(2,19): error TS9003: 'class' expressions are not currently supported.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/anonymousClassExpression1.errors.txt]: No single quotes around "class"

File: tests/baselines/reference/classDeclarationBlockScoping2.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+//// [classDeclarationBlockScoping2.js]
+function f() {
+    var C = (function () {
+        function C() {
+        }
+        return C;
+    })();
+    var c1 = C;
+    {
+        var C = (function () {
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/classDeclarationBlockScoping2.js]: So to be clear, you need to file a bug that we're not correctly renaming these classes.

File: tests/baselines/reference/anonymousClassExpression1.errors.txt (hunk 2 above)
Comment (CyrusNajmabadi) [inline tests/baselines/reference/anonymousClassExpression1.errors.txt]: This matches our naming pattern for the rest of the diagnostics.

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -4551,10 +4552,27 @@ module ts {
...
+                    // We don't need to do this for anything else (like 'extends undefined').  In that
+                    // case you'll crash with an error about being unable to access 'b.prototype'.  
+                    // However, the ES6 spec states that you'd get a TypeError anyways, so it's fine if
+                    // you get an exception from us anyways.
+
                     writeLine();
                     write("var __extends = this.__extends || function (d, b) {");
                     increaseIndent();
                     writeLine();
+                    if (languageVersion === ScriptTarget.ES5) {
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: you need to check this in the checker first. if targetting ES3 it should be an error

Comment (mhegazy) [inline src/compiler/emitter.ts]: We need to discuss this in the design meeting. can you please put it in a separate issue.

I do not think this is related to block scoping anyways.

**Comment by mhegazy** [comment]
The __extends change should be discussed in the design meeting. so please keep this separate.

other than that the change looks good to me.

File: tests/baselines/reference/anonymousClassExpression1.errors.txt (hunk 2 above)
Comment (DanielRosenwasser) [inline tests/baselines/reference/anonymousClassExpression1.errors.txt]: Not really, look where we report errors on function expressions.

**Pull request merged**